File: gateway/src/utils/selector_policy.go
package utils

import (
	"crypto/sha256"
	"encoding/binary"
	commommodel "gateway/src/models/common"
	"math/rand"
)

// PickRoundRobin 返回一个基于轮询的选择结果。
func PickRoundRobin(instances []*commommodel.ServiceInstance, counter int) *commommodel.ServiceInstance {
	if len(instances) == 0 {
		return nil
	}
	return instances[counter%len(instances)]
}

// PickLeastWeightedLatency 返回一个基于最少加权延迟的选择结果。
func PickLeastWeightedLatency(
	instances []*commommodel.ServiceInstance,
	latency map[string]float64,
) *commommodel.ServiceInstance {
	if len(instances) == 0 {
		return nil
	}
	var best *commommodel.ServiceInstance
	var bestScore float64 = -1
	for _, inst := range instances {
		l, ok := latency[inst.ID.String()]
		if !ok {
			l = 0
		}
		score := l / float64(inst.Weight)
		if best == nil || score < bestScore {
			best = inst
			bestScore = score
		}
	}
	return best
}

// PickHashAffinity 返回一个基于哈希亲和性的选择结果。
func PickHashAffinity(instances []*commommodel.ServiceInstance, affinityKey string) *commommodel.ServiceInstance {
	if len(instances) == 0 {
		return nil
	}
	sum := sha256.Sum256([]byte(affinityKey))
	hash := binary.BigEndian.Uint64(sum[:8])
	index := int(hash % uint64(len(instances)))
	return instances[index]
}

// FilterByTags 根据 requireTags 过滤实例列表，返回包含所有 requireTags 的实例。
func FilterByTags(instances []*commommodel.ServiceInstance, requireTags []string) []*commommodel.ServiceInstance {
	if len(requireTags) == 0 {
		return instances
	}
	containsAllTags := func(instTags, reqTags []string) bool {
		tagSet := make(map[string]struct{}, len(instTags))
		for _, t := range instTags {
			tagSet[t] = struct{}{}
		}
		for _, rt := range reqTags {
			if _, ok := tagSet[rt]; !ok {
				return false
			}
		}
		return true
	}
	var result []*commommodel.ServiceInstance
	for _, inst := range instances {
		if containsAllTags(inst.Tags, requireTags) {
			result = append(result, inst)
		}
	}
	return result
}

// RandomWeighted 返回一个基于权重随机选择的实例。
func RandomWeighted(instances []*commommodel.ServiceInstance) *commommodel.ServiceInstance {
	if len(instances) == 0 {
		return nil
	}
	totalWeight := 0
	for _, inst := range instances {
		totalWeight += inst.Weight
	}
	if totalWeight <= 0 {
		return instances[0]
	}
	r := rand.Intn(totalWeight)
	for _, inst := range instances {
		r -= inst.Weight
		if r < 0 {
			return inst
		}
	}
	return instances[len(instances)-1]
}

File: api_service/src/utils/selector_policy.py
import hashlib, random

from src.models.registry.instance import ServiceInstance


def pick_round_robin(
    instances: list[ServiceInstance], counter: int = 0
) -> ServiceInstance | None:
    """轮询选择实例"""
    if not instances:
        return None
    return instances[counter % len(instances)]


def pick_least_weighted_latency(
    instances: list[ServiceInstance], latency_map: dict[str, float]
) -> ServiceInstance | None:
    """选择加权延迟最低的实例"""
    if not instances:
        return None
    ranked = sorted(
        instances, key=lambda i: (latency_map.get(str(i.id), 100.0) / max(i.weight, 1))
    )
    return ranked[0]


def pick_hash_affinity(
    instances: list[ServiceInstance], affinity_key: str
) -> ServiceInstance | None:
    """基于一致性哈希选择实例"""
    if not instances:
        return None
    h = hashlib.sha256(affinity_key.encode()).digest()
    idx = int.from_bytes(h[:4], "big") % len(instances)
    return instances[idx]


def filter_by_tags(
    instances: list[ServiceInstance], require_tags: list[str]
) -> list[ServiceInstance]:
    """根据标签过滤实例"""
    if not require_tags:
        return instances
    req = set(require_tags)
    return [i for i in instances if req.issubset(set(i.tags))]


def random_weighted(instances: list[ServiceInstance]) -> ServiceInstance | None:
    """基于权重随机选择实例"""
    if not instances:
        return None
    total = sum(max(i.weight, 1) for i in instances)
    r = random.uniform(0, total)
    upto = 0
    for i in instances:
        w = max(i.weight, 1)
        if upto + w >= r:
            return i
        upto += w
    return instances[-1]
